File: dmoj/ccc23j5/ccc23j5.cpp
#include <bits/stdc++.h>
using namespace std;

typedef pair<int, int> pii;
#define FR(i, n) for(int i = 0; i < (n); i++)
#define FOR(i, a, b) for(int i = (a); i < (b); i++)
#define FORR(i, a, b) for(int i = (a); i >= (b); i--)

const int MN = 105;
int n, m, w, ans;
string str;
char grid[MN][MN];
int chPos[128];

const pii MOVES[8] = {{-1, -1}, {-1, 0}, {-1, 1}, {0, 1}, {1, 1}, {1, 0}, {1, -1}, {0, -1}};
bool isPref[8], isSuf[8];

void check(int x, int y){
    int pos = chPos[grid[x][y]];
    if(pos == -1) return;

    FR(d, 8){
        auto [dx, dy] = MOVES[d];
        isPref[d] = pos > 0;
        int x1 = x, y1 = y;
        FORR(i, pos-1, 0){
            x1 += dx, y1 += dy;
            if(x1 < 0 || y1 < 0 || x1 >= n || y1 >= m || grid[x1][y1] != str[i]) isPref[d] = false;
        }
        isSuf[d] = pos < w-1;
        x1 = x, y1 = y;
        FOR(i, pos+1, w){
            x1 += dx, y1 += dy;
            if(x1 < 0 || y1 < 0 || x1 >= n || y1 >= m || grid[x1][y1] != str[i]) isSuf[d] = false;
        }
    }

    FR(d, 8){
        ans += isPref[d] && isSuf[(d+2) & 7];
        ans += isPref[d] && isSuf[(d-2) & 7];
        ans += isSuf[d] && pos == 0;
    }
}

int main(){
    cin.sync_with_stdio(0); cin.tie(0);
    cin >> str;
    w = str.size();
    fill(begin(chPos), end(chPos), -1);
    FR(i, w) chPos[str[i]] = i;
    cin >> n >> m;
    FR(i, n) FR(j, m) cin >> grid[i][j];
    FR(i, n) FR(j, m) check(i, j);
    cout << ans << '\n';
}

File: dmoj/ccc23j5/ccc23j5.py
s = input()
n = int(input())
m = int(input())
grid = [list(input().split()) for i in range(n)]

charToPos = [-1]*128
for i in range(len(s)):
    charToPos[ord(s[i])] = i

MOVES = [(-1, -1), (-1, 0), (-1, 1), (0, 1), (1, 1), (1, 0), (1, -1), (0, -1)]
ans = 0

for x in range(n):
    for y in range(m):
        pos = charToPos[ord(grid[x][y])]
        if pos == -1: continue

        isPref = [False]*8
        isSuf = [False]*8

        for d in range(8):
            dx, dy = MOVES[d]

            isPref[d] = pos > 0
            x1, y1 = x, y
            for i in range(pos-1, -1, -1):
                x1 += dx; y1 += dy
                if x1 < 0 or y1 < 0 or x1 >= n or y1 >= m or grid[x1][y1] != s[i]: isPref[d] = False

            isSuf[d] = pos < len(s)-1
            x1, y1 = x, y
            for i in range(pos+1, len(s)):
                x1 += dx; y1 += dy
                if x1 < 0 or y1 < 0 or x1 >= n or y1 >= m or grid[x1][y1] != s[i]: isSuf[d] = False

        for d in range(8):
            ans += isPref[d] and isSuf[(d+2) % 8]
            ans += isPref[d] and isSuf[(d-2) % 8]
            ans += isSuf[d] and pos == 0

print(ans)
